File: src/main/java/br/com/correios/enderecador/dao/RemetenteDao.java
package br.com.correios.enderecador.dao;

import br.com.correios.enderecador.bean.RemetenteBean;
import br.com.correios.enderecador.conexao.ConexaoBD;
import br.com.correios.enderecador.conexao.ConnectException;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;

public class RemetenteDao {
    private final Connection conexao;

    private static RemetenteDao remetenteDao;

    public RemetenteDao() throws DaoException {
        try {
            this.conexao = ConexaoBD.getInstance().recuperaConexao();
        } catch (ConnectException e) {
            throw new DaoException("Não foi possível recuperar a conexão");
        }
    }

    public static RemetenteDao getInstance() throws DaoException {
        if (remetenteDao == null)
            remetenteDao = new RemetenteDao();
        return remetenteDao;
    }

    public void incluirRemetente(RemetenteBean remetenteBean) throws DaoException {
        PreparedStatement stmt = null;
        StringBuilder sql = new StringBuilder();
        try {
            this.conexao.setAutoCommit(true);
            sql.append("INSERT INTO TEC_REMETENTE ");
            sql.append("(REM_TX_APELIDO , REM_TX_TITULO,REM_TX_NOME, ");
            sql.append("REM_ED_CEP, REM_ED_ENDERECO,REM_ED_NUMERO, ");
            sql.append("REM_ED_COMPLEMENTO, REM_ED_BAIRRO, REM_ED_CIDADE, ");
            sql.append("REM_ED_UF, REM_ED_EMAIL, REM_ED_TELEFONE, ");
            sql.append("REM_ED_FAX, REM_ED_CEP_CAIXA_POSTAL, REM_ED_CAIXA_POSTAL) ");
            sql.append("VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ? , ?, ?) ");
            stmt = this.conexao.prepareStatement(sql.toString());
            stmt.setString(1, remetenteBean.getApelido());
            stmt.setString(2, remetenteBean.getTitulo());
            stmt.setString(3, remetenteBean.getNome());
            stmt.setString(4, remetenteBean.getCep());
            stmt.setString(5, remetenteBean.getEndereco());
            stmt.setString(6, remetenteBean.getNumeroEndereco());
            stmt.setString(7, remetenteBean.getComplemento());
            stmt.setString(8, remetenteBean.getBairro());
            stmt.setString(9, remetenteBean.getCidade());
            stmt.setString(10, remetenteBean.getUf());
            stmt.setString(11, remetenteBean.getEmail());
            stmt.setString(12, remetenteBean.getTelefone());
            stmt.setString(13, remetenteBean.getFax());
            stmt.setString(14, remetenteBean.getCepCaixaPostal());
            stmt.setString(15, remetenteBean.getCaixaPostal());
            int cont = stmt.executeUpdate();
            if (cont == 0)
                throw new DaoException("Não foi possível adicionar dados!");
            remetenteBean.setNumeroRemetente(getChave());
        } catch (SQLException e) {
            throw new DaoException(e.getMessage());
        }
    }

    private String getChave() throws DaoException {
        String chave = null;
        PreparedStatement stmt = null;
        StringBuilder sql = new StringBuilder();
        ResultSet rs = null;
        sql.append(" CALL IDENTITY() ");
        try {
            stmt = this.conexao.prepareStatement(sql.toString());
            rs = stmt.executeQuery();
            if (rs.next()) {
                chave = rs.getString(1);
            } else {
                throw new DaoException("Não foi possível recuperar a chave do remetente.");
            }
        } catch (SQLException ex) {
            throw new DaoException(ex.getMessage());
        }
        return chave;
    }

    public void alterarRemetente(RemetenteBean remetenteBean) throws DaoException {
        PreparedStatement stmt;
        StringBuilder sql = new StringBuilder();
        try {
            this.conexao.setAutoCommit(true);
            sql.append("UPDATE TEC_REMETENTE SET ");
            sql.append(" REM_TX_APELIDO=? , REM_TX_TITULO=?,REM_TX_NOME=? ,");
            sql.append(" REM_ED_CEP=?, REM_ED_ENDERECO=?,REM_ED_NUMERO=? ,");
            sql.append(" REM_ED_COMPLEMENTO=?, REM_ED_BAIRRO=?, REM_ED_CIDADE=? ,");
            sql.append(" REM_ED_UF=?, REM_ED_EMAIL=?, REM_ED_TELEFONE=?, ");
            sql.append(" REM_ED_FAX=?, REM_ED_CEP_CAIXA_POSTAL=?, REM_ED_CAIXA_POSTAL=? ");
            sql.append(" WHERE REM_NU= ?  ");
            stmt = this.conexao.prepareStatement(sql.toString());
            stmt.setString(1, remetenteBean.getApelido());
            stmt.setString(2, remetenteBean.getTitulo());
            stmt.setString(3, remetenteBean.getNome());
            stmt.setString(4, remetenteBean.getCep());
            stmt.setString(5, remetenteBean.getEndereco());
            stmt.setString(6, remetenteBean.getNumeroEndereco());
            stmt.setString(7, remetenteBean.getComplemento());
            stmt.setString(8, remetenteBean.getBairro());
            stmt.setString(9, remetenteBean.getCidade());
            stmt.setString(10, remetenteBean.getUf());
            stmt.setString(11, remetenteBean.getEmail());
            stmt.setString(12, remetenteBean.getTelefone());
            stmt.setString(13, remetenteBean.getFax());
            stmt.setString(14, remetenteBean.getCepCaixaPostal());
            stmt.setString(15, remetenteBean.getCaixaPostal());
            stmt.setString(16, remetenteBean.getNumeroRemetente());
            int cont = stmt.executeUpdate();
            if (cont == 0)
                throw new DaoException("Não foi possível alterar dados!");
        } catch (SQLException e) {
            throw new DaoException(e.getMessage());
        }
    }

    public void excluirRemetente(String remetente) throws DaoException {
        PreparedStatement stmt = null;
        StringBuilder sql = new StringBuilder();
        try {
            this.conexao.setAutoCommit(true);
            sql.append("DELETE FROM TEC_REMETENTE WHERE ");
            sql.append("REM_NU = ? ");
            stmt = this.conexao.prepareStatement(sql.toString());
            stmt.setString(1, remetente);
            int cont = stmt.executeUpdate();
            if (cont == 0)
                throw new DaoException("Não foi possível excluir dados!");
        } catch (SQLException e) {
            throw new DaoException(e.getMessage());
        }
    }

    public ArrayList<RemetenteBean> recuperaRemetente(String filtro) throws DaoException {
        PreparedStatement stmt = null;
        ResultSet rs = null;
        ArrayList<RemetenteBean> dados = new ArrayList<>();
        StringBuilder sql = new StringBuilder();
        RemetenteBean remetenteBean = null;
        try {
            sql.append(" SELECT  REM_NU, REM_TX_APELIDO,REM_TX_TITULO,");
            sql.append("         REM_TX_NOME,REM_ED_CEP,REM_ED_ENDERECO,");
            sql.append("         REM_ED_NUMERO,REM_ED_COMPLEMENTO,REM_ED_BAIRRO,");
            sql.append("         REM_ED_CIDADE,REM_ED_UF,REM_ED_EMAIL,REM_ED_TELEFONE,");
            sql.append("         REM_ED_FAX,REM_ED_CEP_CAIXA_POSTAL,REM_ED_CAIXA_POSTAL ");
            sql.append(" FROM  TEC_REMETENTE");
            if (!filtro.equals(""))
                sql.append(filtro);
            sql.append(" ORDER BY UPPER(REM_TX_NOME) ASC ");
            stmt = this.conexao.prepareStatement(sql.toString());
            rs = stmt.executeQuery();
            while (rs.next()) {
                remetenteBean = new RemetenteBean();
                remetenteBean.setNumeroRemetente(rs.getString(1));
                remetenteBean.setApelido(rs.getString(2));
                remetenteBean.setTitulo(rs.getString(3));
                remetenteBean.setNome(rs.getString(4));
                remetenteBean.setCep(rs.getString(5));
                remetenteBean.setEndereco(rs.getString(6));
                remetenteBean.setNumeroEndereco(rs.getString(7));
                remetenteBean.setComplemento(rs.getString(8));
                remetenteBean.setBairro(rs.getString(9));
                remetenteBean.setCidade(rs.getString(10));
                remetenteBean.setUf(rs.getString(11));
                remetenteBean.setEmail(rs.getString(12));
                remetenteBean.setTelefone(rs.getString(13));
                remetenteBean.setFax(rs.getString(14));
                remetenteBean.setCepCaixaPostal(rs.getString(15));
                remetenteBean.setCaixaPostal(rs.getString(16));
                dados.add(remetenteBean);
            }
        } catch (SQLException e) {
            throw new DaoException(e.getMessage());
        }
        return dados;
    }
}

File: src/main/kotlin/br/com/correios/enderecador/dao/RemetenteDao.kt
package br.com.correios.enderecador.dao

import br.com.correios.enderecador.conexao.ConexaoBD
import br.com.correios.enderecador.bean.RemetenteBean
import br.com.correios.enderecador.exception.ConnectException
import br.com.correios.enderecador.exception.DaoException
import org.koin.core.annotation.Singleton
import java.sql.PreparedStatement
import java.lang.StringBuilder
import java.sql.SQLException
import java.sql.Connection

@Singleton
class RemetenteDao(conexaoBD: ConexaoBD) {
    private val conexao: Connection

    init {
        try {
            conexao = conexaoBD.recuperaConexao()
        } catch (e: ConnectException) {
            throw DaoException("Não foi possível recuperar a conexão")
        }
    }

    private fun PreparedStatement.remetenteBean(remetenteBean: RemetenteBean) {
        this.setString(1, remetenteBean.apelido)
        this.setString(2, remetenteBean.titulo)
        this.setString(3, remetenteBean.nome)
        this.setString(4, remetenteBean.cep)
        this.setString(5, remetenteBean.endereco)
        this.setString(6, remetenteBean.numeroEndereco)
        this.setString(7, remetenteBean.complemento)
        this.setString(8, remetenteBean.bairro)
        this.setString(9, remetenteBean.cidade)
        this.setString(10, remetenteBean.uf)
        this.setString(11, remetenteBean.email)
        this.setString(12, remetenteBean.telefone)
        this.setString(13, remetenteBean.fax)
        this.setString(14, remetenteBean.cepCaixaPostal)
        this.setString(15, remetenteBean.caixaPostal)
    }

    fun incluirRemetente(remetenteBean: RemetenteBean) {
        try {
            val stmt = conexao.prepareStatement("""
                INSERT INTO TEC_REMETENTE (
                    REM_TX_APELIDO, REM_TX_TITULO, REM_TX_NOME,
                    REM_ED_CEP, REM_ED_ENDERECO, REM_ED_NUMERO,
                    REM_ED_COMPLEMENTO, REM_ED_BAIRRO, REM_ED_CIDADE,
                    REM_ED_UF, REM_ED_EMAIL, REM_ED_TELEFONE,
                    REM_ED_FAX, REM_ED_CEP_CAIXA_POSTAL, REM_ED_CAIXA_POSTAL) 
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ? , ?, ?)
            """.trimIndent())
            stmt.remetenteBean(remetenteBean)
            stmt.executeUpdate()
            remetenteBean.numeroRemetente = chave
        } catch (e: SQLException) {
            throw DaoException(e.message)
        }
    }

    private val chave: String
        get() {
            val chave: String?
            try {
                val stmt = conexao.prepareStatement(" CALL IDENTITY() ")
                val rs = stmt.executeQuery()
                chave = if (rs.next()) {
                    rs.getString(1)
                } else {
                    throw DaoException("Não foi possível recuperar a chave do remetente.")
                }
            } catch (ex: SQLException) {
                throw DaoException(ex.message)
            }
            return chave
        }

    fun alterarRemetente(remetenteBean: RemetenteBean) {
        try {
            val stmt = conexao.prepareStatement("""
                UPDATE TEC_REMETENTE SET 
                    REM_TX_APELIDO = ?, REM_TX_TITULO = ?, REM_TX_NOME = ?,
                    REM_ED_CEP = ?, REM_ED_ENDERECO = ?, REM_ED_NUMERO = ?,
                    REM_ED_COMPLEMENTO = ?, REM_ED_BAIRRO = ?, REM_ED_CIDADE = ?,
                    REM_ED_UF = ?, REM_ED_EMAIL = ?, REM_ED_TELEFONE = ?,
                    REM_ED_FAX = ?, REM_ED_CEP_CAIXA_POSTAL = ?, REM_ED_CAIXA_POSTAL = ?
                 WHERE REM_NU = ?
            """.trimIndent())
            stmt.remetenteBean(remetenteBean)
            stmt.executeUpdate()
        } catch (e: SQLException) {
            throw DaoException(e.message)
        }
    }

    fun excluirRemetente(remetente: String) {
        try {
            val stmt = conexao.prepareStatement("""
                DELETE FROM TEC_REMETENTE WHERE REM_NU = ?
            """.trimIndent())
            stmt.setString(1, remetente)
            stmt.executeUpdate()
        } catch (e: SQLException) {
            throw DaoException(e.message)
        }
    }

    fun recuperaRemetente(filtro: String = ""): List<RemetenteBean> {
        val dados = mutableListOf<RemetenteBean>()
        try {
            val stmt = conexao.prepareStatement("""
                SELECT  REM_NU, REM_TX_APELIDO, REM_TX_TITULO,
                    REM_TX_NOME, REM_ED_CEP, REM_ED_ENDERECO,
                    REM_ED_NUMERO,REM_ED_COMPLEMENTO,REM_ED_BAIRRO,
                    REM_ED_CIDADE,REM_ED_UF,REM_ED_EMAIL,REM_ED_TELEFONE,
                    REM_ED_FAX,REM_ED_CEP_CAIXA_POSTAL,REM_ED_CAIXA_POSTAL
                    FROM  TEC_REMETENTE
                    $filtro
                    ORDER BY UPPER(REM_TX_NOME) ASC
            """.trimIndent())
            val rs = stmt.executeQuery()
            while (rs.next()) {
                dados.add(RemetenteBean(
                    numeroRemetente = rs.getString(1),
                    apelido = rs.getString(2),
                    titulo = rs.getString(3),
                    nome = rs.getString(4),
                    cep = rs.getString(5),
                    endereco = rs.getString(6),
                    numeroEndereco = rs.getString(7),
                    complemento = rs.getString(8),
                    bairro = rs.getString(9),
                    cidade = rs.getString(10),
                    uf = rs.getString(11),
                    email = rs.getString(12),
                    telefone = rs.getString(13),
                    fax = rs.getString(14),
                    cepCaixaPostal = rs.getString(15),
                    caixaPostal = rs.getString(16),
                ))
            }
        } catch (e: SQLException) {
            throw DaoException(e.message)
        }
        return dados
    }
}
